controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -339,12 +339,32 @@ func (r *BareMetalHostReconciler) credentialsErrorResult(err error, request ctrl
 }
 
 // hasRebootAnnotation checks for existence of reboot annotations and returns true if at least one exist
-func hasRebootAnnotation(host *metal3v1alpha1.BareMetalHost) bool {
-	for annotation := range host.Annotations {
+func hasRebootAnnotation(info *reconcileInfo) (hasReboot, hardMode bool) {
+	for annotation, value := range info.host.GetAnnotations() {
+		// Don't use a break here as we may have multiple clients setting
+		// reboot annotations and we always want hard requests honoured
 		if isRebootAnnotation(annotation) {
-			return true
+			hasReboot = true
+			if isHardReboot(value, info) {
+				hardMode = true
+			}
 		}
 	}
+	return
+}
+
+// isHardReboot checks if reboot annotation requests a hard reboot and returns true if so
+func isHardReboot(annotation string, info *reconcileInfo) bool {
+	annotation_map := map[string]string{}
+	err := json.Unmarshal([]byte(annotation), &annotation_map)
+	if err != nil {
+		info.publishEvent("InvalidAnnotationValue", "reboot annotation has invalid mode, assuming soft-reboot")
```

Comment:
This message would be more helpful if it included the error message from `Unmarshall`, too, like the log message does.

```suggestion
		info.publishEvent("InvalidAnnotationValue", fmt.Sprintf("could not parse reboot annotation (%s), assuming soft reboot mode", err))
```